Introduce dynamic module

CYBAI opened the pull request:
---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #25439
- [x] There are tests for these changes

jdm committed:
checkpoint - dynamic load inside of a module and classic script works.

CYBAI committed:
Update test expectation

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/htmlscriptelement.rs, components/script/dom/servoparser/prefetch.rs, components/script/script_module.rs, components/script/dom/globalscope.rs, components/script/dom/bindings/trace.rs
 * @jgraham: tests/wpt/include.ini
 * @kichjang: components/script/dom/htmlscriptelement.rs, components/script/dom/servoparser/prefetch.rs, components/script/script_module.rs, components/script/dom/globalscope.rs, components/script/dom/bindings/trace.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

CYBAI commented:
This is still WIP so just removed assignee and awaiting review label 🙏

jdm commented on components/script/script_module.rs:
```diff
@@ -882,26 +928,53 @@ impl Callback for ModuleHandler {
     #[allow(dead_code)]
     Worker(TrustedWorkerAddress),
     Window(Trusted<HTMLScriptElement>),
+    DynamicModule(Trusted<DynamicModuleOwner>),
```
I can't figure out how this compiles, given `pub struct Trusted<T: DomObject>` in dom/bindings/refcounted.rs. I am particularly confused because I am clearly compiling your branch that has this line in it!

CYBAI replied:
Because the first field of `DynamicModuleOwner` is `promise` which can be a trusted DomObject 👀 

https://github.com/servo/servo/blob/d5808e0154cc526909799fc4134ba742a3dab906/components/script/script_module.rs#L935-L939

Is this a valid usage of Trusted refcounted thing?

(btw, the `finish_dynamic_module` will be moved to `DynamicModuleOwner` arm in `notify_owner_to_finish` but I was busy this week >_< so haven't tried to fix it; also, I'm wondering it might be good to have a trait to define the `notify_owner_to_finish` function? so that we can just call the function in `ModuleOwner::notify_owner_to_finish`)

jdm replied:
Oh, I completely overlooked the `#[dom_struct]` on DynamicModuleOwner. We require `TrustedPromise` instead of `Trusted<Promise>`, and we should consider some other way of holding the DynamicModuleOwner. I honestly don't know if the current implementation is expected to work or not.

CYBAI replied:
I just figured out why we're hitting TIMEOUT for some tests and why https://austineng.github.io/webgpu-samples/ also gets the TIMEOUT.

Basically, the root cause is,

when the owner is `ModuleOwner::Script`, we will send the IPC route to `CoreResourceMsg::Fetch` via the `document_loader::fetch_async`; however, when the owner is a dynamic loader (or a worker in the future), nothing will handle the IPC router so it will just be dropped on the floor. Then, no one knows the route should be sent.

https://github.com/servo/servo/blob/d5808e0154cc526909799fc4134ba742a3dab906/components/script/script_module.rs#L1708-L1717

So, we should handle the IPC router much correctly for non-script cases.

then, while I tried to experimentally send the IPC action

```rust
match document {
    Some(doc) => doc.fetch_async(LoadType::Script(url.clone()), request, action_sender),
    None => {
        let _ = global
            .resource_threads()
            .sender()
            .send(CoreResourceMsg::Fetch(
                request,
                FetchChannels::ResponseMsg(action_sender, None),
            ))
            .unwrap();
    },
}
```

> I honestly don't know if the current implementation is expected to work or not.

I just got the answer of this and it's `No, it won't work`.

jdm replied:
Well spotted!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #27084) made this pull request unmergeable. Please resolve the merge conflicts.

CYBAI committed:
Introduce DynamicModuleOwner dom struct

CYBAI commented:
@jdm @kunalmohan Good news here!

### About WebGPU

In current `dynamic-module` branch, I can't see the red triangle though, the https://austineng.github.io/webgpu-samples can be loaded successfully! (I remember I even can't scroll in previous impl) and if I click in the code block (which looks like a squashed code block), then I will get `window.focus is not a function` error.

```
→ ./mach run --pref=dom.webgpu.enabled https://austineng.github.io/webgpu-samples/\#helloTriangle
[2020-07-04T09:01:50Z ERROR script::dom::bindings::error] Error at https://austineng.github.io/webgpu-samples/dist/common-606af6.js:1:197397 window.focus is not a function
[2020-07-04T09:01:53Z ERROR script::dom::bindings::error] Error at https://austineng.github.io/webgpu-samples/dist/common-606af6.js:1:197397 window.focus is not a function
```

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/6782666/86509161-e5b07880-be20-11ea-856a-782d0d2563cb.png">

### About Dynamic Import

I'm not sure if it's a good idea but I wonder it should be safer and easier to generate the `DynamicModuleOwner` from codegen? With using a codegen-ed version DynamicModuleOwner, codegen has already handled how to GC for us so we can just use it in `script_module`! The `DynamicModuleOwner` webidl is introduced in cc2a9e7 commit.

Then, with simply sending the IPC router to resource threads by `global.resource_threads().sender()` in f2e73b6, it has already fixed most of TIMEOUT!

But, there are still some unexpected FAIL and TIMEOUT! I'm going to observe the root cause for remaining FAIL and TIMEOUT! 

Hope current branch can work fine for the webgpu project! 🙏

kunalmohan commented:
Awesome! Thank you @cybai!!

>I can't see the red triangle though

I wonder why that happens :thinking: . I'll try to investigate this.

CYBAI commented:
> > I can't see the red triangle though
> 
> I wonder why that happens 🤔 . I'll try to investigate this.

It looks like the module is loaded but if you think it _might_ be something wrong in dynamic module, feel free to let me know! 😄

kunalmohan commented:
cc @kvark

bors-servo commented:
:umbrella: The latest upstream changes (presumably #27088) made this pull request unmergeable. Please resolve the merge conflicts.

kvark commented:
@kunalmohan if you get stuck investigating this, please consider exposing a way for Servo to record the wgpu API trace. We'd be able to look at it outside of Servo and see if it captures the problem.

CYBAI committed:
Move from direct `Evaluate` to `compile then execute` for classic script

CYBAI committed:
Set private reference for classic script

Web developers can use `Dynamic Import` in a classic script; thus, we
need to save the script's private reference so that we can reuse it when
we're going to fetch a dynamic import module for a classic script.

Besides, because it's possible to use different executing context for a
dynamic import module (like `dynamic-import/string-compilation-other-document.html` WPT test),
we can't initialize a module owner at the timing of `SetScriptPrivate`;
thus, if the private module script doesn't hold an owner, we'll use a
DynamicImport owner for it.

CYBAI commented:
@bors-servo try=wpt

- Show me if the `compile then execute` breaks anything!

bors-servo commented:
:hourglass: Trying commit 84efc1e27656a96bc7d9cb4eb5855bbcbd1a6c5f with merge 52aa3f48534a2d00f514781fa32025e9768c8764...

CYBAI commented on tests/wpt/metadata/html/semantics/scripting-1/the-script-element/module/dynamic-import/no-active-script-manual-module.html.ini:
```diff
@@ -0,0 +1,2 @@
+[no-active-script-manual-module.html]
+  expected: TIMEOUT
```
These 2 tests are introduced in web-platform-tests/wpt#14157 for testing "no active script" triggered by user click; however, there's no way to trigger an user click in WPT for now (see web-platform-tests/wpt#14158) so they were expected TIMEOUT.

I was disappointed that we always need to wait for a few seconds for these 2 expected TIMEOUT tests but, luckily, Domenic changed filename for these 2 files to match the manual test naming convention in web-platform-tests/wpt#24565 so they will be seen as manual tests and won't be run in CI.

I also checked them locally and yes, both of them are PASS!

(I will remove both of this expected TIMEOUT when next WPT sync are merged!)

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/GDnL2GsLSO6g95_AArPoDg)

CYBAI commented:
Wow, many unexpected CRASH and ERROR; looks like I do break something 😂 Let me investigate it!

CYBAI commented:
@bors-servo try=wpt

- Does the workaround help? If yes, then I might have some questions.

bors-servo commented:
:hourglass: Trying commit 959750d572be2725d1f6b2e13a7fe87952407db1 with merge 1c4c15af0a59d9f59ed22a797bfbbade0b1fc324...

jdm commented:
You likely will want to use the Compile function instead of CompileForNonSyntacticScope, since the compiled script isn't being executed with a specific scope chain (eg https://searchfox.org/mozilla-central/rev/ae81637193d6687398669169f9808000a96996ee/dom/base/nsJSUtils.cpp#219-221).

jdm commented:
However it will be important to null-check the result of the compilation like https://searchfox.org/mozilla-central/rev/ae81637193d6687398669169f9808000a96996ee/dom/base/nsJSUtils.cpp#222-226.

CYBAI commented:
> You likely will want to use the Compile function instead of CompileForNonSyntacticScope, since the compiled script isn't being executed with a specific scope chain (eg https://searchfox.org/mozilla-central/rev/ae81637193d6687398669169f9808000a96996ee/dom/base/nsJSUtils.cpp#219-221).

Ahhh! Yes! Thanks for catching it so quick! I understood it in a wrong way!

> However it will be important to null-check the result of the compilation like https://searchfox.org/mozilla-central/rev/ae81637193d6687398669169f9808000a96996ee/dom/base/nsJSUtils.cpp#222-226.

Yes! I just noticed I forgot the null-check; with a checking, I can fix unexpected failure of `/html/webappapis/scripting/processing-model-2/window-onerror-parse-error.html`.

But, both `Compile` and `CompileForNonSyntacticScope` return a `JSScript`; I found I can't use `is_null` to check if the returned js script object is a null object or not (so I just used `JS_IsExceptionPending` to prove my assessment of missing null-check).

I checked mozjs but didn't find a way to convert JSScript into Value which supports `is_null`.
Is it possible to make `Compile` return an `Option`? Or, is it possible to make `JSScript` support a `is_null` function?

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/azxqOdEiSaaNvMaQbXXGZg)

jdm commented:
The return value is `*mut JSObject`, so it should be possible to call https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.is_null on it.

CYBAI commented:
> The return value is `*mut JSObject`, so it should be possible to call https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.is_null on it.

Ahhhhh, no, I was wrong 😭 I didn't use `is_null`; I used `is_undefined` to check :| how can I read the document with `null` and check it with `undefined` 🤦‍♂️

But one more question here 🙋‍♀️

Does every mozjs objects imply `JSObject`? With checking the doc of [`Compile1`](https://doc.servo.org/mozjs/jsapi/fn.Compile1.html), it says the functions will return a `JSScript` but I just can use the `is_null` function. Or, I should say, because of `*mut JSScript`, it's a pointer so I can use the function from primitive pointer in Rust? 🤔 

Thanks btw!

CYBAI commented:
@bors-servo try=wpt

- Use the correct null-check instead of workaround!

bors-servo commented:
:hourglass: Trying commit 8b0466042db97226eb47678d435fb9e968707b7d with merge 314df221d1dab943f3c20f5aa88bd791e7738013...

jdm commented:
Right, the is_null that I suggested is implemented for all raw pointer types by the Rust standard library. It's unrelated to the concept of a null JS object.

jdm commented:
And I see why my previous comment was confusing; I intended to write that the return value was `*mut JSScript`, not `*mut JSObject`.

CYBAI commented:
> Right, the is_null that I suggested is implemented for all raw pointer types by the Rust standard library. It's unrelated to the concept of a null JS object.

I see! Thank you!

I think the `tests/wpt/web-platform-tests/html/webappapis/dynamic-markup-insertion/document-write/module-dynamic-import.html` is the last unexpected FAIL I need to investigate!

Will try to investigate it and update more information for the expected FAIL in this PR!

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/dp4YJOT8RYebT4boqRav0w)

CYBAI commented on tests/wpt/metadata/html/semantics/scripting-1/the-script-element/module/dynamic-import/string-compilation-other-document.html.ini:
```diff
@@ -1,16 +1,3 @@
-    expected: FAIL
-
   [setTimeout should successfully import]
     expected: FAIL
```
For `setTimeout` triggered by string code, we currently failed at finding its correct execution context so it will fetch a wrong link.

Corresponding spec link: https://html.spec.whatwg.org/multipage/#timer-initialisation-steps

I haven't digged into it but I think we need to pass base url of the timer and we might need to save it in `JsTimerTask`.

I wonder if it's good to be a follow-up issue, if not, I can try to fix it in this PR!

(Interestingly, only Firefox executes correctly on all the `setTimeout` tests for dynamic import module 😂)

jdm replied:
I did a bit of reading in Gecko:
* they get the active script reference when setTimeout is invoked (https://searchfox.org/mozilla-central/rev/89814940895946b48b4c04c702efd2c676ec8e7e/dom/base/nsGlobalWindowInner.cpp#5718)
* this calls JS_GetScriptedCallerPrivate, which retrieves the private value from the JSScript that is currently running
* when it's time to execute the timer, they compile it and set the private value of the new JSScript to point to the active script reference that invoked setTimeout (https://searchfox.org/mozilla-central/rev/89814940895946b48b4c04c702efd2c676ec8e7e/dom/base/nsGlobalWindowInner.cpp#5749-5756)

CYBAI replied:
Thanks for the investigation! Sounds like we can set the private reference of the script and then get it when setTimeout is invoked so that we can get a proper base_url 👀

CYBAI replied:
Filed #27260

CYBAI commented on tests/wpt/metadata/html/semantics/scripting-1/the-script-element/module/dynamic-import/string-compilation-nonce-module.html.ini:
```diff
@@ -1,19 +1,7 @@
-    expected: FAIL
-
   [reflected inline event handlers must not inherit the nonce from the triggering script, thus fail]
     expected: FAIL
```
We don't support `nonce` yet; same as `dynamic-import/string-compilation-nonce-classic.html.ini`.

CYBAI commented on tests/wpt/metadata/html/semantics/scripting-1/the-script-element/module/dynamic-import/dynamic-imports-script-error.html.ini:
```diff
@@ -1,32 +1,4 @@
-    expected: FAIL
-
   [import() must reject with different error objects for each import when there is a instantiation error]
     expected: FAIL
```
This is a bug in current module script implementation because I misunderstood the spec.

Instead of saving the error and always using it, we need to return a different instantiation error if it's imported from different script element or `dynamic import`.

I think refactoring this might make this PR harder to review so I'd like to file a follow-up issue for it and fix it separately!

CYBAI replied:
Filed #27261

CYBAI commented:
@jdm this is ready for review now! I just commented for the test expectation and hope it helps for reviewing!

Also, for https://github.com/servo/servo/pull/27026#discussion_r453334558 and https://github.com/servo/servo/pull/27026#discussion_r453334250, these two might be potential follow-up issues. If it's good to fix them later, please let me know and I will file issues for them and fix them later! Thanks! 🙏

CYBAI commented on components/script/script_module.rs:
```diff
@@ -871,19 +872,22 @@ pub(crate) enum ModuleOwner {
     pub fn notify_owner_to_finish(&self, module_identity: ModuleIdentity) {
         match &self {
             ModuleOwner::Worker(_) => unimplemented!(),
+            ModuleOwner::DynamicModule(_) => unimplemented!(),
```
Currently, we will register `finish_dynamic_module` with `append_dynamic_module_handler` which is different from the normal `notify_owner_to_finish` flow so this is `unimplemented` for now.

But, I will want to refactor it to be a _trait_ and hold a `notify_to_finish` function so that we can just call it here and then I can try to remove the `append_dynamic_module_handler` function to remove more duplicated-like code!

jdm commented:
The suggested follow up issues sound sensible to me!

jdm commented on components/script/script_module.rs:
```diff
@@ -880,6 +940,74 @@ impl ModuleOwner {
+        let existing_rethrow_error = module_tree.get_rethrow_error().borrow().as_ref().cloned();
+
+        let execution_err = if network_error.is_none() && existing_rethrow_error.is_none() {
+            if let Some(ref record) = &*module_tree.get_record().borrow() {
+                let evaluated = module_tree.execute_module(&global, record.handle()).err();
```
I think it might be risky to keep the borrow active while we call execute_module. We should clone the record instead.

CYBAI replied:
Indeed! Let me fix it! Thanks!

CYBAI replied:
Fixed in fe0462d

jdm reviewed (changes requested):
These changes look sensible! I haven't come up with a better solution than the DynamicModuleOwner WebIDL interface; we don't have any other examples of structs containing SpiderMonkey values that need to be passed around async tasks that aren't already DOM objects, so let's stick with the solution that works.

jdm commented on components/script/dom/globalscope.rs:
```diff
@@ -2561,26 +2570,58 @@ impl GlobalScope {
+
+                        SetScriptPrivate(
+                            *compiled_script,
+                            &PrivateValue(Box::into_raw(module_script_data) as *const _),
```
After reading the full PR, we do not call Box::from_raw. We will need to use [SetScriptPrivateReferenceHooks](https://searchfox.org/mozilla-central/rev/89814940895946b48b4c04c702efd2c676ec8e7e/js/src/jsapi.h#1793-1798) - I believe if we use Rc instead of Box (and Rc::into_raw and Rc::from_raw), we can call clone() on the value in the addref hook, and drop the value in the release hook, and it should work ok. See https://searchfox.org/mozilla-central/rev/89814940895946b48b4c04c702efd2c676ec8e7e/dom/script/LoadedScript.cpp#70-88 for Gecko's implementation.

CYBAI replied:
Thanks for commenting on this! I planned to ask about it but forgot >_<

To make sure I understand correctly,

With the functionality of reference counting, when we use `Rc` instead of `Box` and then call `clone()` in `addref` hook, the counting will be added automatically, is that correct?

and a question here 👇 

1. For dropping values in release hook, I'm not clear how can I do that? Should I explicitly call `drop(private_reference)`?

btw, maybe it's good to handle it in this PR? I imagine the private references won't be released if we don't do the drop in the release hook; we're making all scripts (both classic and module) to own an private reference so it might be good to release them properly?

jdm replied:
Yes, if we do `let val = RC::from_raw(...); mem::forget(val.clone()); mem::forget(val);` in the addref hook, that will add one reference to the value. In the release hook, we do `let _val = Rc::from_raw(...);` and it will automatically subtract one reference when we return.

CYBAI replied:
Fixed in 731a16e!

I basically copied the code in previous comment to add ref hook but I'm not sure if I clearly understand how the `mem::forget` works here.

Does that mean

```rust
let val = Rc::from_raw(value); // 1. Get the existing script reference from Rc raw pointer
mem::forget(val.clone());        // 2. Increment the Rc count by doing a `clone` but release it without running drop
mem::forget(val);                    // 3. Release the existing value without running drop
```

With `mem::forget`, we can release the memory here but still make it counted for the Rc pointer; is that correct?

CYBAI commented on tests/wpt/metadata/html/webappapis/dynamic-markup-insertion/document-write/module-dynamic-import.html.ini:
```diff
@@ -1,5 +1,4 @@
-  expected: TIMEOUT
   [document.write in an imported module]
-    expected: TIMEOUT
+    expected: FAIL
```
oh, indeed! hmmm.. then I'm really curious why this one failed :/ we do fire the load event though...

I will file a follow-up issue for it, too! Thanks!

CYBAI replied:
Filed #27262

CYBAI commented on components/script/dom/globalscope.rs:
```diff
@@ -2529,7 +2537,7 @@ impl GlobalScope {
     /// Evaluate JS code on this global scope.
     pub fn evaluate_js_on_global_with_result(&self, code: &str, rval: MutableHandleValue) -> bool {
-        self.evaluate_script_on_global_with_result(code, "", rval, 1)
+        self.evaluate_script_on_global_with_result(code, "", rval, 1, None)
```
Yes! I think it would be good to always pass a URL but one of the reason why I made it an Option is, if an user tries to import a module from devtools, it might not have a URL?

Or, let's just use `global.api_base_url()` in that case 👀? (which might make more sense!)

jdm replied:
I'm inclined to make the argument non-optional and make every caller decide.

CYBAI replied:
I see! Taking non-optional makes sense to me! I will fix it! Thanks!

CYBAI replied:
It's fixed in 870692c 

(Also see https://github.com/servo/servo/pull/27026#discussion_r454778062)

CYBAI commented on components/script/dom/globalscope.rs:
```diff
@@ -2561,26 +2570,58 @@ impl GlobalScope {
+
+                        let module_script_data = Box::new(ModuleScript::new(
+                            base_url,
+                            ScriptFetchOptions::default_classic_script(&self),
```
I also noticed maybe we should use its host fetch options. Does it make sense to make `HTMLScriptElement` hold a `ScriptFetchOptions` so that it can be reused here when trying to evaluate the script?

Also, for cases like devtools, should we accept an Option here and fallback to default or just accept a `ScriptFetchOptions` 👀?

jdm replied:
I think storing it in HTMLScriotElement might be awkward, since it will either be an Option or a fake default value for part of it's existence. If we can store it in Script origin and ScriptContext while the fetching is happening, then we can create the ModuleScript with the real value. Does that make sense?

CYBAI replied:
> If we can store it in Script origin and ScriptContext while the fetching is happening, then we can create the ModuleScript with the real value.

ah! yes! this makes sense to me! I will try to fix it in this way! thanks!

CYBAI replied:
Hmm, while trying to add options to ScriptOrigin, I notice ScriptOrigin derives JSTraceable and MallocSizeOf traits which means we will need to make all properties of ScriptFetchOptions to derive these 2 traits?

If we try to pass `options` to all the `*_loaded` functions and pass it to the `execute` function, then we can reuse the options in `run_a_classic_script`. Does that make sense?

jdm replied:
If that makes it easier, then that sounds fine.

CYBAI replied:
hmmm... I found I also need to update a bunch of places in the codebase with passing options; maybe deriving the traits would be easier 😂

CYBAI replied:
hmm, no, while I saw this compilation error, I think it might be better to use the passing options way because I believe we want to use `JSTraceable` internally inside the script crate?

```error: cannot find derive macro `JSTraceable` in this scope
  --> components/net_traits/request.rs:37:37
   |
37 | #[derive(Clone, Debug, Deserialize, JSTraceable, MallocSizeOf, PartialEq, Serialize)]
   |                                     ^^^^^^^^^^^
```

Will use the passing options way to fix this!

jdm replied:
The other alternative is to add a new entry with `unsafe_no_jsmanaged_fields!` in components/script/dom/bindings/trace.rs.

CYBAI replied:
I fixed this with 3 separated commits.

- 870692c0aa55 : Make `script_base_url` non-optional
- 8c1eba6cdbd8 : Make `ScriptOrigin` to hold its fetch options
- f300f22396aa58 : Use fetch options from ScriptOrigin

Hope this will make review easier 🙏

CYBAI replied:
Btw, using `unsafe_no_jsmanaged_fields!` in this case makes sense to me because I think we might not need to trace those options and it works fine! Thanks for the tip!

CYBAI commented on components/script/dom/webidls/DynamicModuleOwner.webidl:
```diff
@@ -0,0 +1,13 @@
+ * so that we can hold a traceable owner for those dynamic modules which don't hold a owner.
+ */
+
+[Exposed=(Window)]
```
Thanks for the tip! This makes sense to me why we can revert the change for `mozilla/interfaces` test!

CYBAI replied:
Fixed it in the original commit which is 950c159 now.

CYBAI commented on components/script/script_module.rs:
```diff
@@ -1066,8 +1196,160 @@ pub unsafe fn EnsureModuleHooksInitialized(rt: *mut JSRuntime) {
+        Destination::Script,
+        options,
+        None,
+        //parent_url, XXXjdm
```
@jdm thanks for the question! I think I should remove this comment.

Let me explain why 👇 

#### Should we provide a parent identity for dynamic imports?

I think No

#### Why don't we need a parent identity?

Having a parent identity for a module means the module is a sub-module of the parent identity.

In the case of dynamic import, a dynamic module might be imported in a sub-module but itself is not a sub-module of the module.

So, I think we don't need to provide a parent identity in this case!

(Hope this explanation makes sense! If not, please let me know!)

CYBAI replied:
Removed the comment in 7bafb94

CYBAI commented on components/script/script_module.rs:
```diff
@@ -1169,25 +1450,71 @@ pub fn fetch_external_module_script(
+    requests: Vec<RootedTraceableBox<DynamicModule>>,
+
+    #[ignore_malloc_size_of = "Define in uuid"]
+    next_id: DynamicModuleId,
```
Fixed in f334952

CYBAI commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -13959,7 +13959,7 @@
     ],
     "interfaces.html": [
-     "d988b7939f83cef16a51a4ae5502ed029f31df99",
+     "fbad40c9d97d674a149f50a2cff4155574decd05",
```
Dropped the commit.

CYBAI commented on tests/wpt/mozilla/tests/mozilla/interfaces.html:
```diff
@@ -74,6 +74,7 @@
   "DOMStringList",
   "DOMStringMap",
   "DOMTokenList",
+  "DynamicModuleOwner",
```
Dropped the commit.

CYBAI commented:
Other than https://github.com/servo/servo/pull/27026#discussion_r453849053, I've fixed most comments with fixup commits. To have an easier squash step in the future, I just use `commit --fixup` so I leave comments for each comments above! Hope it won't make review harder!

Also, I will think about https://github.com/servo/servo/pull/27026#discussion_r453818187 (if we need `Option<Dom<Document>>` for dynamic module) later.

CYBAI commented on components/script/script_module.rs:
```diff
@@ -1252,17 +1596,17 @@ pub fn fetch_single_module_script(
 
     let document: Option<DomRoot<Document>> = match &owner {
-        ModuleOwner::Worker(_) => None,
+        ModuleOwner::Worker(_) | ModuleOwner::DynamicModule(_) => None,
```
After doing some investigation, I think no.

Let me list 2 points why we don't need to store an optional document for a dynamic module owner:

#### Explanation of my investigation

1. It's possible to use `Dynamic Import` in a worker context.
2. The usage of this `document` is to queue the IPC into document loader which is specifically for `<script type=module>`. For dynamic import module, I think we need to understand it in a JS way which means browser engine part only needs to help handle fetch the dynamic module, compile it, link it and so on. For example, when we try to dynamically import a module inside a classic script, the `import()` syntax will return a promise; thus, IIUC, for a browser engine, we only need to handle the [`HostImportModuleDynamically`](https://html.spec.whatwg.org/multipage/#hostimportmoduledynamically(referencingscriptormodule%2C-specifier%2C-promisecapability)) properly.

For the 1st point, I found there are dynamic import tests for workers under [/workers/modules/dedicated-worker-import.any.js](https://wpt.fyi/results/workers/modules/dedicated-worker-import.any.worker.html). We don't enable the module workers tests yet but I found we're able to pass the dynamic import tests in that file. So, I think my hypothesis might be right? Also, according to [baku's comment in Bugzilla 1540913](https://bugzilla.mozilla.org/show_bug.cgi?id=1540913#c8), I can confirm it's correct that worker will also support dynamic import but Firefox side need to refactor a bit to support this feature.

#### Conclusion

If above investigation makes sense, then we don't need to make the dynamic module owner to hold an `Option<Dom<Document>>`.

CYBAI committed:
Handle script private reference hooks properly

jdm commented on components/script/dom/workletglobalscope.rs:
```diff
@@ -93,6 +94,7 @@ impl WorkletGlobalScope {
         self.globalscope.evaluate_js_on_global_with_result(
             &*script,
             rval.handle_mut(),
+            ScriptFetchOptions::default_classic_script(&self.globalscope),
```
Please file a followup issue to take a ScriptFetchOptions argument so we can make it match https://github.com/servo/servo/blob/854cc793861c31c395387a1197c40e840bc95826/components/script/dom/worklet.rs#L627-L634.

CYBAI replied:
Thanks for the reminder! I just filed it at #27308!

jdm commented on components/script/timers.rs:
```diff
@@ -545,6 +546,7 @@ impl JsTimerTask {
                 global.evaluate_js_on_global_with_result(
                     code_str,
                     rval.handle_mut(),
+                    ScriptFetchOptions::default_classic_script(&global),
```
Please note in #27260 that we will want to use the fetch options from the original script as well.

CYBAI replied:
Commented!

jdm commented on components/script/script_module.rs:
```diff
@@ -1214,10 +1216,26 @@ pub unsafe fn EnsureModuleHooksInitialized(rt: *mut JSRuntime) {
 
+#[allow(unsafe_code)]
+unsafe extern "C" fn host_add_ref_top_level_script(value: *const Value) {
+    let val = Rc::from_raw(value);
```
We need to use `value.to_private() as *const ModuleScript` here.

CYBAI replied:
ok! 🙆‍♂️ Thanks for the tip!

But, I'm don't clearly understand why we need to do the `to_private` and `as *const ModuleScript` here? Does that mean using `to_private` and cast to a raw pointer of `ModuleScript` is correctly pointed to the same one we saved before?

jdm replied:
Correct.

CYBAI replied:
I see! Thanks! Always happy to learn more :D

CYBAI replied:
hmm, wait, this commit doesn't compile! Let me fix it!

jdm commented on components/script/script_module.rs:
```diff
@@ -1214,10 +1216,26 @@ pub unsafe fn EnsureModuleHooksInitialized(rt: *mut JSRuntime) {
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn host_release_top_level_script(value: *const Value) {
+    let _val = Rc::from_raw(value);
```
We need to do the same cast here as well.

jdm reviewed (changes requested):
This is very close!

CYBAI commented on components/script/script_module.rs:
```diff
@@ -1243,9 +1256,9 @@ pub unsafe extern "C" fn host_import_module_dynamically(
-#[derive(Clone)]
+#[derive(Clone, MallocSizeOf)]
 /// <https://html.spec.whatwg.org/multipage/#script-fetch-options>
-pub(crate) struct ScriptFetchOptions {
```
While using this type as parameter types in other modules, rustc will complain about `can't leak crate-visible type` so I just removed the crate only visiable keyword, `(crate)`, here 👀

CYBAI commented on components/script/dom/bindings/trace.rs:
```diff
@@ -200,6 +201,8 @@ unsafe_no_jsmanaged_fields!(BroadcastChannelRouterId);
 unsafe_no_jsmanaged_fields!(BlobId);
 unsafe_no_jsmanaged_fields!(BlobImpl);
 
+unsafe_no_jsmanaged_fields!(ScriptFetchOptions);
```
Fixed in  f929fdf !

jdm commented on components/script/script_module.rs:
```diff
@@ -1226,14 +1226,14 @@ pub unsafe fn EnsureModuleHooksInitialized(rt: *mut JSRuntime) {
 #[allow(unsafe_code)]
 unsafe extern "C" fn host_add_ref_top_level_script(value: *const Value) {
-    let val = Rc::from_raw(value);
+    let val = Rc::from_raw(value.to_private() as *const ModuleScript);
```
Does this build? I expect you need to do `(*value).to_private()`, since it's a pointer.

CYBAI replied:
> Does this build? I expect you need to do `(*value).to_private()`, since it's a pointer.

No! I just fixed it with what you expected locally but still under building!

CYBAI replied:
Fixed again in  5830cf5 >_<

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 5830cf54637d656231ff7acc3aa72c2d34576c1e with merge 1031fbafaa9e76cea34b2264b9db0bac46cce47e...

jdm commented:
If this passes tests, I think it's ready to land after squashing!

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/GBHxI5dKRkOXEyAJXGgeYQ)

CYBAI commented:
That is #25522 so I'm going to rebase and squash! Thanks!

CYBAI commented:
@bors-servo try=wpt

- In case of breaking anything while rebase, let me try it again! (specifically for the `Rc<DOMString>` `ScriptOrigin` `text`)

Btw, @jdm does the squashed commits make sense for you? If not, I'm also fine to squash all of them into one commit 👀

bors-servo commented:
:hourglass: Trying commit 08723b0e37a83c705810a17f8923882689171345 with merge c7ad2cc41776357efdd8e968b90434ef2b2ae11e...

jdm commented:
Yep, that looks sensible. Thanks!

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/UXV9Abk9S6Ofk8u7xbHI6g)
State: approved= try=True

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 08723b0 has been approved by `jdm`

<!-- @bors-servo r=jdm 08723b0e37a83c705810a17f8923882689171345 -->

bors-servo commented:
:hourglass: Testing commit 08723b0e37a83c705810a17f8923882689171345 with merge ee88827a21577a0fd157726fb7269266b8a78959...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/GaiozX2vRU-ReejwSa1Asw)

jdm commented:
@bors-servo retry
* #27311

bors-servo commented:
:hourglass: Testing commit 08723b0e37a83c705810a17f8923882689171345 with merge 6f354b026669e401f7a3c91bd8d55e7be790c274...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/QJQ4Wt-1SVOv0_8TIau6Tw)

jdm commented:
Huh, #27311 hit us again. I'm getting suspicious, so I'll try building this branch on my Windows machine and seeing if I can reproduce it.

jdm commented:
I can reproduce the crash consistently.

jdm commented:
Somehow, we're calling GetScriptPrivate on a null JSScript, despite checking for a null pointer first. That's... confusing.
```
servo!js::BaseScript::sourceObject+0x3
servo!JS::GetScriptPrivate+0x3
servo!script::dom::globalscope::{{impl}}::evaluate_script_on_global_with_result::{{closure}}+0x5f4
servo!profile_traits::time::profile<bool,closure-0>+0x13f
servo!script::dom::globalscope::GlobalScope::evaluate_script_on_global_with_result+0x222
servo!script::dom::htmlscriptelement::HTMLScriptElement::run_a_classic_script+0x383
servo!script::dom::htmlscriptelement::HTMLScriptElement::execute+0x717
servo!script::dom::htmlscriptelement::HTMLScriptElement::prepare+0x1d73
servo!script::dom::servoparser::ServoParser::tokenize<closure-0>+0x244
servo!script::dom::servoparser::ServoParser::do_parse_sync+0x289
servo!script::dom::servoparser::{{impl}}::parse_sync::{{closure}}+0x16
servo!profile_traits::time::profile<(),closure-0>+0xf7
servo!script::dom::servoparser::ServoParser::parse_sync+0x1ab
servo!script::dom::servoparser::ServoParser::parse_bytes_chunk+0xa9
servo!script::dom::servoparser::{{impl}}::process_response_chunk+0x11e
servo!script::script_thread::ScriptThread::handle_fetch_chunk+0x10b
servo!script::script_thread::ScriptThread::handle_msg_from_constellation+0x2b8
servo!script::script_thread::{{impl}}::handle_msgs::{{closure}}+0x13a
servo!script::script_thread::ScriptThread::profile_event<closure-5,core::option::Option<bool>>+0xcf
servo!script::script_thread::ScriptThread::handle_msgs+0x2507
```

CYBAI commented:
Hmm, is it possible that it means we might have a bug in GetScriptPrivate on Windows 🤔?

jdm commented:
Ugh, you're absolutely right. Great catch!

jdm commented:
More specifically, it's our old nemesis "can't return JSValue by value on Windows". We'll need to add a method to jsapi::glue that takes a `*mut JSValue` like https://github.com/servo/rust-mozjs/blob/0caf5549cddbb34ad16abf35fb6bfbff824a4d14/src/jsglue.cpp#L1095-L1098.

CYBAI commented:
> More specifically, it's our old nemesis "can't return JSValue by value on Windows".

Oh, TIL 🤯 Thanks!

> We'll need to add a method to jsapi::glue that takes a `*mut JSValue` like https://github.com/servo/rust-mozjs/blob/0caf5549cddbb34ad16abf35fb6bfbff824a4d14/src/jsglue.cpp#L1095-L1098.

Thanks! Let me try to send a PR!

CYBAI committed:
Use JS_GetScriptPrivate to get a mutated return value

Because MSVC uses a different calling conventions for functions that
return non-POD values, we need to use the new exposed wrapper function
so that `GetScriptPrivate` can be handled correctly on Windows.

CYBAI commented:
@bors-servo try

- I confirmed it works fine in my local though, I'd like to see if it also works fine in both Linux and Windows (it should fix the ANGLE smoketest on windows and have no impact to linux)

bors-servo commented:
:hourglass: Trying commit 419cd53561e4abaadc78b2cfea55e0bf0edfb36c with merge a9cb7377d9823b27f594e2d4e2ab0a103fe09276...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/XpgOpzVuRrC2xBgDQr-lGg)
State: approved= try=True

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 419cd53 has been approved by `jdm`

<!-- @bors-servo r=jdm 419cd53561e4abaadc78b2cfea55e0bf0edfb36c -->

bors-servo commented:
:hourglass: Testing commit 419cd53561e4abaadc78b2cfea55e0bf0edfb36c with merge 086556e706bbb65790e7fb9a918bf7e8051171e0...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/KOmSBHhsReKaGKTUHvyDRg)
Approved by: jdm
Pushing 086556e706bbb65790e7fb9a918bf7e8051171e0 to master...
